syntax = "proto3";

option go_package = "intf/jointrpc";

service JointRPC {
  rpc Call(JSONRPCCallRequest) returns (JSONRPCCallResult);
  rpc Notify(JSONRPCNotifyRequest) returns (JSONRPCNotifyResponse);

  rpc ListMethods(ListMethodsRequest) returns (ListMethodsResponse);
  rpc ListDelegates(ListDelegatesRequest) returns (ListDelegatesResponse);

  // state stream
  rpc SubscribeState(AuthRequest) returns (stream SubscribeStateResponse);
  
  // request/response dual streams
  rpc Worker(stream JSONRPCEnvolope) returns (stream JSONRPCEnvolope);
}

message Empty {}
message Status {
  int32 code = 1;
  string reason = 2;
}

message ClientAuth {
  string username = 1;
  string password = 2;
}

message JSONRPCEnvolope {
  string body = 1;  // json encoded rpc message
  string trace_id = 2;
}

message ListMethodsRequest {
  ClientAuth auth = 1;
}

message ListMethodsResponse {
  Status status = 1;
  repeated MethodInfo methods = 2;
}

message ListDelegatesRequest {
  ClientAuth auth = 1;
}

message ListDelegatesResponse {
  Status status = 1;
  repeated string delegates = 2;
}

// message DeclareMethodsRequest {
//   string request_id = 1;  
//   repeated MethodInfo methods = 2;

// }

// message DeclareMethodsResponse {
//   string request_id = 1;  
//   Status status = 2;
// }

// message DeclareDelegatesRequest {
//   string request_id = 1;  
//   repeated string methods = 2;
// }

// message DeclareDelegatesResponse {
//   string request_id = 1;  
//   Status status = 2;

// }

message JSONRPCCallRequest {
  ClientAuth auth = 1;
  JSONRPCEnvolope envolope = 2;
  bool  broadcast = 3;
  int32   timeout = 4;     // the seconds from expire this rpc
}


message JSONRPCCallResult {
  Status status = 1;
  JSONRPCEnvolope envolope = 2;
}

message JSONRPCNotifyRequest {
  ClientAuth  auth = 1;
  JSONRPCEnvolope envolope = 2;
  bool broadcast = 3;
}

message JSONRPCNotifyResponse {
  Status status = 1;
  string text = 2;
}

message MethodInfo {
  string name = 1;
  string help = 2;
  string schema_json = 3;
}

message ServerState {
  repeated MethodInfo methods = 1;
}

message AuthRequest {
  string request_id = 1;  
  ClientAuth client_auth = 2;
}

message AuthResponse {
  string request_id = 1;  
  Status status = 2;
  string namespace = 3;
}

message SubscribeStateResponse {
  oneof payload {
     AuthResponse auth_response = 1;
     ServerState state = 2;
  }
}
